Registry update

rstoneback opened the pull request:
# Description

Addresses #516 and more

Updates the register package.
 - Changed format of .pysat/user_modules.txt to store packages by platform and name
 - Changed format of pysat.user_modules to be a dict of dicts keyed by platform and name
 - Restricted `pysat_*` instruments to those internal to pysat.instruments
 - Changed use of registry so that only requested module needs to be imported
 - Added support to add multiple packages at once using strings
 - Added support to register an entire instrument package like `pysat.instruments` or `pysatModels.models`
 - Added support to remove all instrument modules from registry by platform name along
 - Added support to remove multiple packages by string platform and name
 - Modified method calls and docstrings
 - Raises error if a module is removed but not registered
 - Raises error if a module is already registered for a given platform and name
 - Expanded unit tests

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

Unit tests as well as through review of a pysatSpaceWeather instrument, ACE.

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

rstoneback commented:
To Do:
 - Use pysat testing infrastructure to ensure modules input for registration satisfy all of pysat's minimum requirements
 - Update changelog
 - Update documentation

rstoneback commented on pysat/utils/registry.py:
```diff
@@ -84,48 +111,154 @@ def register(module_name):
-    registry.register('my.package.name.myInstrument')
-    assert 'my.package.name.myInstrument' in user_modules
+        registry.register('my.package.name.myInstrument')
+        assert 'my.package.name.myInstrument' in user_modules
```
This needs an update.

rstoneback commented on pysat/utils/registry.py:
```diff
@@ -84,48 +111,154 @@ def register(module_name):
+            raise
+
+        # second, check that module is itself pysat compatible
+
```
This is where the minimum feature set for a pysat instrument needs to go.

rstoneback replied:
pinging @jklenzing since you are closest to the updated test structure.

rstoneback replied:
A couple options are in the code

rstoneback replied:
If there are no other opinions I prefer the second option.

 Also thinking we can write a method in utils, something like verify_pysat_standard (compatibility?). Input a instrument module, get a report on how it fares against pysat standards, and test against methods. Basically a simple way to test things in development.

aburrell replied:
> A couple options are in the code

It may be more useful for discussion to put them in the conversation here as examples or put them on a slide/wiki page as examples to be discussed at the next meeting.

rstoneback replied:
Good point.

prep work in containing method.
```
inst_module = importlib.import_module(module_name)      
validate = itc.InstTestClass()
```

First option, using the test class as is.
```
        class Foo(object):
            pass
        # parse string to get package part and instrument module part
        parse = module_name.split('.')

        validate.package = Foo()
        validate.package.__name__ = '.'.join(parse[:-1])
        validate.test_modules_standard(parse[-1])
        validate.test_standard_function_presence(parse[-1])
```
No modifications to the testing code is required. Requires a dummy class and relation of prep code to use by validate isn't clear unless familiar with validate.

Using the test class with a slight mod for outside use. Doesn't impact use of current test code.
```
        validate.test_modules_standard('', inst_module)
        validate.test_standard_function_presence('', inst_module)
```
I like the direct input of the module under test, though it comes with the downside of having a '' leading input and changes to input of test methods.

rstoneback replied:
I went ahead and cleaned things up in favor of the first option and using the code as it exists. Not worth adding in an optional code path for something there is already have a solution for. I spent a bit of time on the comments instead.

rstoneback commented:
I'll take a pass at documentation once #524 is resolved.

rstoneback commented:
I also need to pull in the Lock, Network Lock once that is resolved as well in #376 and #525

aburrell commented on pysat/__init__.py:
```diff
@@ -79,20 +79,29 @@
     # user modules file
     with open(os.path.join(pysat_dir, 'user_modules.txt'), 'w') as f:
         f.write('')
```
```suggestion
    with open(os.path.join(pysat_dir, 'user_modules.txt'), 'w') as fout:
        fout.write('')
```

aburrell commented on pysat/__init__.py:
```diff
@@ -79,20 +79,29 @@
 
 else:
     # load up stored data path
     with open(os.path.join(pysat_dir, 'data_path.txt'), 'r') as f:
         data_dir = f.readline()
```
```suggestion
    with open(os.path.join(pysat_dir, 'data_path.txt'), 'r') as fin:
        data_dir = fin.readline()
```

aburrell commented on pysat/__init__.py:
```diff
@@ -79,20 +79,29 @@
+                    user_modules[platform][name] = inst_module
     else:
         # write user modules file
         with open(os.path.join(pysat_dir, 'user_modules.txt'), 'w') as f:
```
Fix single character variable name here and elsewhere.

rstoneback replied:
Those changes were in #376, and now #525, and in progress.

aburrell commented on pysat/tests/test_registry.py:
```diff
@@ -41,53 +44,346 @@ def create_fake_module(full_module_name, platform, name):
+
+
+class TestModuleRegistration():
+
```
Shouldn't this have a setup and teardown?  I thought we settled on that at the last meeting.

rstoneback replied:
Fixed.

aburrell commented on pysat/utils/registry.py:
```diff
@@ -15,117 +15,296 @@
 
 
-def remove(*module_names):
+def remove(platforms, names=None):
```
Wouldn't it be cleaner to have `names` default to an empty list?  This would make the expected input clearer and not affect the functionality.

aburrell replied:
I provided a suggestion in the docstring, but didn't make changes elsewhere.  It's really just the `is None` evaluation line that would need to be changed to `len(names) == 0` and also add a ValueError check for `len(names) != len(platforms)`.  This should probably be done anyhow.

rstoneback replied:
General guidance I've read is that objects that are modifiable in place shouldn't be used as defaults in the method signature. If names=[], or names='hi there!', is in the call signature and names gets modified in-place within the method, those changes will persist between calls since the signature definition code is only run at definition.

while it seems contrary to the whole point of the feature, I've seen the best-practices method signature is
```
def new_method(var1, keyword=None):
    if keyword is None:
        keyword = []
```

aburrell replied:
>  objects that are modifiable in place shouldn't be used as defaults in the method signature

Yes, I am ok with `names` being an arg instead of a kwarg.  Is there any instance where someone wouldn't be expected to enter both?  It's required in every other area.

General guidance also recommends avoiding the use of `isinstance` for testing your input.  You can get around this by using `np.asarray` on the input.  This will be more robust in instances where someone has an input that for some reason isn't interpreted as a string (as has happened before and will likely happen again).

rstoneback replied:
>> Is there any instance where someone wouldn't be expected to enter both? It's required in every other area.

If a user wants to remove all instruments for a given platform(s). I like having a minimal number of required inputs but I am ok with changing names to an arg. I'll have something tomorrow.

 I do like to be able to support a single string or list-like of strings with the same code. I've tried using np.asarray and it hasn't worked out for me so far.
```
In [19]: for item in np.asarray('hello'): 
    ...:     print(item) 
    ...:                                                                                                                                                              
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-bff22eb1af27> in <module>
----> 1 for item in np.asarray('hello'):
      2     print(item)
      3 

TypeError: iteration over a 0-d array

```

I've tried duck typing multiple times on the string vs list issue. I've also tried checking for iter or other things like that. None are quite right so far.

rstoneback replied:
Switched to list-like for args platforms and names. Similar change to other methods that allowed a string or list like of strings, now just list-like.

aburrell replied:
Cool.  For future reference, this was an alternative:
```
platforms = np.asarray(platforms)

if platforms.shape == ():
    platforms = np.array([platforms])
```

rstoneback replied:
Cool. I figured the registry wasn't a high user interaction area so probably not worth the potential future maintenance costs of fancy inputs. I'll remember this for areas that will see more use though!

aburrell commented on pysat/utils/registry.py:
```diff
@@ -15,117 +15,296 @@
-
-    Notes
+    ----------
+    platforms : str of list-like of str
```
```suggestion
    platforms : str or list-like of str
```

aburrell commented on pysat/utils/registry.py:
```diff
@@ -15,117 +15,296 @@
+    ----------
+    platforms : str of list-like of str
+        One or more platform identifiers to remove
+    names : str of list-like of str
+        Name identifiers, paired with platforms, to remove.
+        If names is None, then all instruments under
+        `platforms` will be removed. Supports a mixed
+        combination of name labels and None. (default=None)
```
```suggestion
    names : str or list-like of str
        Name identifiers, paired with platforms, to remove.
        If the names element paired with the platform element is None,
        then all instruments under the specified platform will be
        removed. Should be the same type as `platforms`. An empty list
        will default to removing all instruments for all specified
        platforms. (default=[])
```

rstoneback replied:
I adopted the suggestion apart from the default.

aburrell commented on pysat/tests/test_registry.py:
```diff
@@ -1,9 +1,12 @@
 import pytest
 import sys
 
+import pysat
 from pysat import user_modules
 from pysat.utils import registry
 from pysat import Instrument
```
Since you're now importing pysat, is it really useful to separately importing `user_modules` and `Instrument`?

aburrell commented on pysat/utils/registry.py:
```diff
@@ -15,117 +15,305 @@
+        # first, ensure module string directs to something importable
+        try:
+            inst_module = importlib.import_module(module_name)
+        except Exception:
```
Do we know what type of Exception is expected?

rstoneback replied:
My intent with this try: except is verify we can import the supplied input and provide some context to any error if one occurs. The raise a couple lines below preserves the original error and stack trace for the user to deal with.  

No errors are allowed on an import to ensure, to the extent possible, that the registry only has good packages written to it.

rstoneback replied:
There use to be more of these in earlier versions. It started with my logging the error in a variety of places, since, logging is good. But most of the entries would end up like,
```
pysat ERROR: 'platname1'
```
which wasn't all that helpful. I tried a few versions but non seemed to really fit the bill. This one was a residual. After thinking about my response above, it really doesn't do very much.

rstoneback replied:
I also have a similar modification over in _instrument.py, so worth sorting out our style.
```
                # not a native pysat.Instrument
                try:
                    mod = user_modules[self.platform][self.name]
                    inst = importlib.import_module(mod)
                except Exception:
                    estr = ' '.join(('pysat was either unable to locate the',
                                     'module in the registry or',
                                     'successfully import the module',
                                     'for platform:', self.platform,
                                     'and name:', self.name))
                    logger.error(estr)
                    raise

```

rstoneback replied:
The notion of friendlier error message came up on a recent review that threw a KeyError when I forgot a tag.

aburrell replied:
Well, the reason I don't like this is that there are too many things it will catch.  I like catching exceptions and giving friendlier messages, but specifying which one would be more useful.  For example, you could have:
```
try:
    mod = user_modules[self.platform][self.name]
    inst = importlib.import_module(mod)
except KeyError as kerr:
    estr = "unknown platform or name supplied to user modules: {:} or {:} not in {:}".format(self.platform, self.name, user_modules.__repr__())
    logger.error(estr)
    raise KeyError(kerr)
except ImportError as ierr:
    estr = "unable to locate or import module for platform {:}, name {:}".format(self.platform, self.name)
    logger.error(estr)
    raise ImportError(ierr)
```

Then, if something else breaks we'll still get it in the intended form and can fix or catch it as appropriate.

rstoneback commented:
Review comments have been incorporated. Ready for review @aburrell. Cheers.

aburrell commented on pysat/tests/test_registry.py:
```diff
@@ -36,58 +40,385 @@ def create_fake_module(full_module_name, platform, name):
+                        zip(module_strings, self.platforms,
+                            self.platform_names)]
+
+        # remove any existing support which may be let over
```
```suggestion
        # remove any existing support which may be left over
```
?

aburrell commented on pysat/tests/test_registry.py:
```diff
@@ -36,58 +40,385 @@ def create_fake_module(full_module_name, platform, name):
+        # remove any existing support which may be let over
+        # errors if you try to remove a package that isn't
+        # registered
+        for plat_name in np.unique(self.platforms):
+            try:
+                registry.remove([plat_name], [None])
+            except ValueError:
+                pass
```
Testing may benefit from raising a warning for platforms that can't be removed.

rstoneback replied:
The ValueError was raised if someone tried to remove a platform that wasn't registered. I removed that error.

 logger.info now gets a message that the package wasn't registered.

rstoneback commented:
I modified the pull based upon conversation on Wed. Errors raised are reduced. Added an overwrite flag to register. 
ValueError raised by .register if someone tries to register a module that has a platform and name combo already registered for a different module, and overwrite set to False. Same module is ok. Existing registry may be updated with overwrite set to True.
